File: backend/app/schemas/article.py
from datetime import datetime
from typing import Annotated, Optional
from pydantic import BaseModel, Field, field_validator
from models.article import ArticleViewEnum # TODO move these enums here?
from models.category import *
from schemas.category_preview import CategoryPreview
import schemas.user as UserSchema
import re

INVALID_URL_PATTERN = re.compile(r"[^\w]", re.A) # Catches non-alphanumeric characters (including non-ASCII), except underscore.

class ArticleBase(BaseModel):
    """Base schema class with common validators."""
    @field_validator("filename", check_fields=False)
    def validate_url(cls, filename: str):
        if INVALID_URL_PATTERN.search(filename): # Cannot contain slashes or characters that are reserved or would require url-encoding
            raise ValueError("Invalid filename")
        return filename

    @field_validator("content", check_fields=False)
    def validate_content(cls, content: str):
        if len(content) == 0:
            raise ValueError("Content must be not empty")
        return content
    
class ArticleInput(ArticleBase):
    """Schema for creation requests."""
    filename: Annotated[str, Field(description="Identifier. Must be unique within the article's category. Does not require an extension.")]
    title: str
    content: Annotated[str, Field(description="Markdown-like text content.")]
    text: Annotated[str, Field(description="Raw unformatted text transcript of the article.")]
    summary: Annotated[str, Field(description="Should be unformatted text.")]

class ArticleUpdate(ArticleBase):
    """Schema for patching requests."""
    filename: Optional[str] = None
    title: Optional[str] = None
    content: Optional[str] = None # Raw document text
    is_draft: Optional[bool] = None # If true, the content will be saved as a draft and not published
    featured_image_path: Optional[str] = None
    text: Optional[str] = None
    publish_time: Optional[str] = None # As ISO 8601 date
    is_visible: Optional[bool] = None
    view_type: Optional[ArticleViewEnum] = None
    can_comment: Optional[bool] = None
    show_authors: Optional[bool] = None
    category_sorting_index: Optional[int] = None
    authors: Optional[list[str]] = None # List of usernames
    category_path: Optional[str] = None
    summary: Optional[str] = None
    tags: Optional[list[str]] = None

class ArticlePreview(ArticleBase):
    """Schema for basic article metadata, without content fields."""
    id: int
    filename: str
    title: str
    creation_time: datetime
    publish_time: Optional[datetime]
    is_visible: bool
    category_path: str
    category_id: int
    category_name: str
    path: str # Full path to article
    category_sorting_index: int
    authors: list[UserSchema.UserOutput]
    summary: str
    tags: list[str]
    can_comment: bool
    comments_count: int
    featured_image_path: Optional[str]

class ArticleOutput(ArticlePreview):
    """Complete article schema that includes metadata and content."""
    category: CategoryPreview
    content: str # Raw document text
    view_type: ArticleViewEnum
    last_edit_time: Optional[datetime]
    show_authors: bool

class ArticleSearchResults(BaseModel):
    results: list[ArticlePreview]

class ArticleLatestPosts(ArticleSearchResults):
    total_articles: int

class TagOutput(BaseModel):
    """Schema for a tag entity."""
    id: int
    name: str

class TagsOutput(BaseModel):
    """Schema for tags used in articles on the site."""
    tags: list[TagOutput]

File: frontend/services/article.ts
/**
 * Utility methods for /articles/ API routes.
 */
import type { CategoryPreview } from './category'
import Service from './service'
import type { User } from './user'

export type articleID = integer
export type ArticleViewMode = "single_page"|"by_sections"

export type ArticlePreview = {
  id: integer,
  filename: string,
  title: string,
  creation_time: dateISOString,
  publish_time?: dateISOString,
  is_visible: boolean,
  category_path: string,
  category_id: integer,
  category_name: string,
  /** Full path to article */
  path: string,
  category_sorting_index: integer,
  authors: User[],
  summary: string,
  tags: string[],
  can_comment: boolean,
  comments_count: integer,
  featured_image_path: path,
}

export type Article = ArticlePreview & {
  category: CategoryPreview,
  /** Raw document text */
  content: string,
  view_type: ArticleViewMode,
  show_authors: boolean,
  last_edit_time: dateISOString,
  parent_category_names: string[],
  annotations: ArticleAnnotation[]
}

export type ArticleCreationRequest = {
  filename: string,
  title: string,
  /** Raw document text */
  content: string,
  summary: string,
  text: string,
}

export type ArticleAnnotationInput = {
  id: integer,
  author: string,
  comment: string,
  start: integer,
  end: integer,
}

export type ArticleAnnotation = {
  id: integer,
  author: User,
  comment: string,
  start: integer,
  end: integer,
}

export type ArticleUpdateRequest = {
  filename?: string,
  title?: string,
  /** Markdown document */
  content?: string,
  publish_time?: dateISOString,
  is_visible?: boolean,
  view_type?: "single_page"|"by_sections",
  can_comment?: boolean,
  show_authors?: boolean,
  category_sorting_index?: integer,
  authors?: string[],
  category_path?: string,
  summary?: string,
  /** Plaintext transcript of the article */
  text?: string,
  tags?: string[],
  is_draft?: boolean,
  featured_image_path?: path,
  annotations?: ArticleAnnotationInput[],
}

export type Tag = {
  id: integer,
  name: string
}

export type Tags = {
  tags: Tag[]
}

class ArticleService extends Service {
  constructor(api_url: string) {
    super(api_url)
  }

  /** Creates an article. */
  async createArticle(path: path, request: ArticleCreationRequest): Promise<Article> {
    const response = await this.post('/articles' + path, request)
    return response.data
  }

  /** Fetches an article by its path; expects a leading slash. */
  async getArticle(path: path, draft?: boolean): Promise<Article> {
    const response = await this.get('/articles' + path, {
      draft: draft,
    })
    return response.data
  }

  /** Updates an article's data. */
  async updateArticle(path: path, request: ArticleUpdateRequest): Promise<Article> {
    const response = await this.patch('/articles' + path, request)
    return response.data
  }

  /** Fetches all tags that have been used on the site. */
  async getAllTags(): Promise<Tags> {
    return (await this.get('/articles/tags')).data
  }
}

export default ArticleService